File: features/ffi/wrapper/Foldl.hs
module Foldl() where

import Foreign
import Foreign.C.Types

foreign import ccall unsafe "foldl.h foldl" c_foldl ::
	FunPtr (CInt -> CInt -> CInt) -> CInt -> CInt -> Ptr CInt -> CInt

foreign import ccall unsafe "foldl.h &add" c_add :: FunPtr (CInt -> CInt -> CInt)
foreign import ccall unsafe "foldl.h &mul" c_mul :: FunPtr (CInt -> CInt -> CInt)

foreign import ccall unsafe "foldl.h &sample" c_sample :: Ptr CInt

foreign import ccall unsafe "wrapper" wrap ::
	(CInt -> CInt -> CInt) -> IO (FunPtr (CInt -> CInt -> CInt))

File: features/ffi/wrapper/c/foldl.c
typedef int (*Op)(int x, int y);

int
foldl(Op op, int x0, int n, int xs[])
{
	int s = x0; int i = 0;

	for (i = 0; i < n; i ++) s = op(s, xs[i]);
	return s;
}

int add(int x, int y) { return x + y; }
int mul(int x, int y) { return x * y; }

int sample[] = { 3, 5, 8, 9 };
